spec.bs

```diff
@@ -309,18 +308,121 @@ The following are the [=event handlers=] (and their corresponding [=event handle
+</dl>
+
+<div algorithm>
+  The <dfn method for="AppHistory">navigate(|url|, |options|)</dfn> method steps are:
+
+  1. <a spec="HTML" lt="parse a URL">Parse</a> |url| relative to [=this=]'s [=relevant settings object=]. If that returns failure, then return [=a promise rejected with=] a "{{SyntaxError}}" {{DOMException}}. Otherwise, let |urlRecord| be the <a spec="HTML">resulting URL record</a>.
+
+  1. Return the result of [=performing an app history navigation=] given [=this=], |urlRecord|, and |options|.
+</div>
+
+<div algorithm>
+  The <dfn method for="AppHistory">navigate(|options|)</dfn> method steps are:
+
+  1. If neither |options|["{{AppHistoryNavigationOptions/navigateInfo}}"] nor |options|["{{AppHistoryNavigateOptions/state}}"] [=map/exists=], then return [=a promise rejected with=] a {{TypeError}}.
+
+  1. Set |options|["{{AppHistoryNavigateOptions/replace}}"] to true.
+
+     <p class="note">This is not technically necessary, as the main <a spec="HTML">navigate</a> algorithm will override any non-"<a for="history handling behavior">`replace`</a>" history handling behavior for same-URL navigations.
+
+  1. Let |urlRecord| be [=this=]'s [=relevant global object=]'s [=active document=]'s [=Document/URL=].
+
+  1. Return the result of [=performing an app history navigation=] given [=this=], |urlRecord|, and |options|.
+</div>
+
+<div algorithm>
+  To <dfn>perform an app history navigation</dfn> given an {{AppHistory}} object |appHistory|, a [=URL=] |url|, and an {{AppHistoryNavigateOptions}} |options|:
+
+  1. Let |browsingContext| be |appHistory|'s [=relevant global object=]'s [=Window/browsing context=].
+
+  1. Let |historyHandling| be "<a for="history handling behavior">`replace`</a>" if |options|["{{AppHistoryNavigateOptions/replace}}"] [=map/exists=] and is true; otherwise, "<a for="history handling behavior">`default`</a>".
+
+  1. Let |navigateInfo| be |options|["{{AppHistoryNavigationOptions/navigateInfo}}"] if it exists; otherwise, undefined.
+
+  1. Let |serializedState| be null.
+
+  1. If |options|["{{AppHistoryNavigateOptions/state}}"] [=map/exists=], then set |serializedState| to [$StructuredSerializeForStorage$](|options|["{{AppHistoryNavigateOptions/state}}"]). If this throws an exception, return [=a promise rejected with=] that exception.
+
+  1. Let |promise| be [=a new promise=] created in |appHistory|'s [=relevant Realm=].
+
+  1. Let |previousPromise| be |appHistory|'s [=AppHistory/navigate method call promise=].
+
+  1. Set |appHistory|'s [=AppHistory/navigate method call promise=] to |promise|.
+
+  1. <a spec="HTML">Navigate</a> |browsingContext| to |url| with <i>[=navigate/historyHandling=]</i> set to |historyHandling|, <i>[=navigate/appHistoryInfo=]</i> set to |navigateInfo|, <i>[=navigate/appHistoryState=]</i> set to |serializedState|, and the <a spec="HTML">source browsing context</a> set to |browsingContext|.
+
+  1. Set |appHistory|'s [=AppHistory/navigate method call promise=] to |previousPromise|.
```

Comment:
Hmm, I'll need to think more about this. It just seemed like good practice to reset it back. Let me try working through some examples.

(In simple cases this'll just reset it back to null, but in nested cases it might matter.)

Reply:
I worked through the following example:

```js
appHistory.addEventListener("navigate", e => {
  if (e.destination === url1) {
    const p2 = appHistory.navigate(url2);
  }
  e.respondWith(Promise.resolve());
});
const p1 = appHistory.navigate(url1);
```

With this spec draft `p2` and `p1` both fulfill. The URL and history update steps run first for url2 and then for url1 (because `respondWith()` happens for url2 first). This kind of makes sense. But it's not great.

I think we should tackle this at a higher level. How about the following: any call to the navigate algorithm (not just `appHistory.navigate()`, but any navigation) while a `navigate` event is ongoing forcibly cancels the ongoing navigate event, so that the new one can proceed unmolested. Probably we should also do the URL and history update steps algorithm so that `history.pushState()` during `navigate` has the same impact.

Then for the above example `p1` would reject with an "`AbortError`" `DOMException` and `p2` would fulfill. And the URL would transition to `url2` and never hit `url1`.

I'm still unsure where this leaves the original question as to whether we should restore previousPromise or not, but if this direction sounds good I can work it out.

---

A related question: consider a tweak of the above

```js
appHistory.addEventListener("navigate", e => {
  e.respondWith(Promise.resolve());
  if (e.destination === url1) {
    const p2 = appHistory.navigate(url2);
  }
});
const p1 = appHistory.navigate(url1);
```

Here the developer has already called respondWith(). At least with the current semantics this means the url will update to url1. But probably we should fix that per https://github.com/WICG/app-history/issues/94 's "Change the updating of location.href and appHistory.current to synchronously happen after all handlers have finished running".